### ISSUE TITLE
Question about : v2.2.4.3 verified as compliant with the CTT  

### ISSUE THREAD

**Issue by Askidea** [iotivity/iotivity-lite#321]
Hello,

I have a few questions regarding the CTT certification.

I saw that v2.2.4.3 was verified to be CTT compatible in tags description. If so, I would like to know in what test environment it was performed at the time of verification. The CTT version will be 2103.0.0, I wonder what the IUT is and what the PICS file contents are.

I am developing OCF Client in the Android app to which the iotivity library is applied, and when I test with any iotivity library version with CTT 2103, several FAIL items occur for all.
Also, depending on the build options (ALL=1) or (CLOUD=0 TCP=0 others=1), the success or failure of the CTT test item may vary slightly.

To summarize the questions:
1) At the time of testing with CTT 2103, what were the IUT and PICS?
2) How did you specify the build options for the iotivity library version?
3) Have you tested it in both Linux and Android environments?

I would be grateful if someone who knows well about the above could give me some answers. I hope you know that this is very important to me and urgent.

Thanks

**Comment by SiMet** [comment]
Hi,

We test it only with Linux env.
We use CTT from lastest code (As a Comarch we develop this tool), so it could contain some not released fixes and updates.

We use For testing we use few sample applications:

ServerCertification (TCP=1 IPV4=0 SWUPDATE=1 CREATE=1 MNT=1 CLOUD=1 DEBUG=0 V6DNS=1 WKCORE=1)
PICS:```{
    "device": "Device_name",
    "company": "Company_name",
    "platformVersion": "1.0",
    "role": "Server",
    "icv": "ocf.2.2.5",
    "dmv": "ocf.res.1.3.0, ocf.sh.1.3.0",
    "supportedVerticalProfile": [
        "Smart Home"
    ],
    "supportedDeviceTypes": [
        "oic.d.switch"
    ],
    "resources": [
        "oic.r.sdi",
        "oic.wk.mnt",
        "oic.r.switch.binary",
        "oic.r.sp",
        "oic.wk.res",
        "oic.wk.con",
        "oic.r.csr",
        "oic.r.roles",
        "oic.r.temperature",
        "oic.wk.col",
        "oic.r.ael",
        "oic.r.coapcloudconf",
        "oic.r.softwareupdate",
        "oic.r.dali",
        "oic.r.dali.conf",
        "oic.r.remotecontrol"
    ],
    "optionalFeatures": [
        "batchInterface",
        "supportsWellknownCoreDiscovery"
    ],
    "OTM": [
        "oic.sec.doxm.jw",
        "oic.sec.doxm.rdp",
        "oic.sec.doxm.mfgcert"
    ],
    "contentFormatVersion": [
        "1.0.0"
    ],
    "acceptVersion": [
        "1.0.0"
    ],
    "supportedSecurityProfiles": [
        "sp-baseline-v0",
        "sp-black-v0",
        "sp-blue-v0",
        "sp-purple-v0"
    ],
    "cplAtIANAPen": "1.3.6.1.4.1.71",
    "cplAtModel": "Discovery",
    "cplAtVersion": "1.0",
    "observableOICRES": false,
    "sct": 457,
    "persistentDeviceuuid": false,
    "RDPublishCapability": true,
    "SSIDSoftAP": "",
    "PasswordSoftAP": "",
    "canTurnOnSoftApManually": true,
    "jurisdictionSwitch": false,
    "minNotifyPeriod": [
        0,
        10000
    ],
    "maxNotifyPeriod": [
        0,
        20000
    ],
    "threshold": [
        0,
        5
    ],
    "threshold_recommended_value": 2,
    "multiValueQuerySupport": true,
    "atomicMeasurements": [
        
    ],
    "supportsCoAPoverTCP": true,
    "OIC11Support": false,
    "supportsMultipleDTLS": true
}```
OnboardingTool (TCP=1 IPV4=0 SWUPDATE=1 CREATE=1 MNT=1 CLOUD=1 DEBUG=0)
PICS:
```{
    "device": "OBT",
    "company": "Intel Corporation",
    "platformVersion": "1.0",
    "role": "Client",
    "icv": "ocf.2.2.5",
    "dmv": "ocf.res.1.3.0",
    "supportedVerticalProfile": [
        "Smart Home"
    ],
    "supportedDeviceTypes": [
        "oic.d.dots",
        "oic.d.cms",
        "oic.d.ams",
        "oic.wk.d",
        "oic.d.airpurifier"
    ],
    "resources": [
        "oic.r.csr",
        "oic.r.roles",
        "oic.r.sp",
        "oic.wk.mnt",
        "oic.d.dots",
        "oic.d.cms",
        "oic.d.ams",
        "oic.r.softwareupdate"
    ],
    "optionalFeatures": [
        
    ],
    "OTM": [
        "oic.sec.doxm.jw",
        "oic.sec.doxm.rdp",
        "oic.sec.doxm.mfgcert"
    ],
    "contentFormatVersion": [
        "1.0.0"
    ],
    "acceptVersion": [
        "1.0.0"
    ],
    "supportedSecurityProfiles": [
        "sp-baseline-v0",
        "sp-black-v0",
        "sp-blue-v0",
        "sp-purple-v0"
    ],
    "cplAtIANAPen": "1.3.6.1.4.1.71",
    "cplAtModel": "Discovery",
    "cplAtVersion": "1.0",
    "multiValueQuerySupport": false,
    "sct": 1,
    "supportsCoAPoverTCP": true,
    "canTurnOnSoftAPManually": false,
    "observableOICRES": false,
    "supportsCancellingObserve": false,
    "persistentDeviceuuid": false,
    "onboardingRoles": [
        "ams",
        "cms",
        "dots"
    ],
    "onboardingOTM": [
        "oic.sec.doxm.jw",
        "oic.sec.doxm.rdp",
        "oic.sec.doxm.mfgcert"
    ],
    "onboardingSpec": [
        "ocf.1.3"
    ]
}```
ClientCertification (TCP=1 IPV4=0 SWUPDATE=1 CREATE=1 MNT=1 DEBUG=0 V6DNS=1 WKCORE=1)
PICS:
```{
    "device": "Client",
    "company": "Company_name",
    "platformVersion": "1.0",
    "role": "Client",
    "icv": "ocf.2.2.5",
    "dmv": "ocf.res.1.3.0, ocf.sh.1.3.0",
    "supportedVerticalProfile": [
        "Smart Home"
    ],
    "supportedDeviceTypes": [
        "oic.d.light"
    ],
    "resources": [
        "oic.r.sdi",
        "oic.wk.res",
        "oic.r.csr",
        "oic.r.ael",
        "oic.r.sp",
        "oic.r.roles",
        "oic.wk.mnt",
        "oic.wk.con",
        "oic.r.softwareupdate"
    ],
    "optionalFeatures": [
        "batchInterface",
        "ipv6Scope3",
        "ipv6Scope5",
        "supportsWellknownCoreDiscovery"
    ],
    "OTM": [
        "oic.sec.doxm.jw",
        "oic.sec.doxm.rdp",
        "oic.sec.doxm.mfgcert"
    ],
    "contentFormatVersion": [
        "1.0.0"
    ],
    "acceptVersion": [
        "1.0.0"
    ],
    "supportedSecurityProfiles": [
        "sp-baseline-v0",
        "sp-black-v0",
        "sp-blue-v0",
        "sp-purple-v0"
    ],
    "cplAtIANAPen": "1.3.6.1.4.1.71",
    "cplAtModel": "Discovery",
    "cplAtVersion": "1.0",
    "multiValueQuerySupport": false,
    "sct": 457,
    "supportsCoAPoverTCP": true,
    "canTurnOnSoftApManually": false,
    "observableOICRES": true,
    "supportsCancellingObserve": true,
    "persistentDeviceuuid": false,
    "onboardingRoles": [
        
    ],
    "onboardingOTM": [
        
    ],
    "onboardingSpec": [
        
    ],
    "PasswordSoftAP": null,
    "RDPublishCapability": false,
    "SSIDSoftAP": null,
    "atomicMeasurements": [
        
    ],
    "authorizationServerUrl": null,
    "cis": null,
    "cloudAccountPassword": null,
    "cloudAccountUrl": null,
    "cloudAccountUser": null,
    "cloudClientCertificate": null,
    "cloudClientPrivateKey": null,
    "cloudServerTrustAnchorCertificate": null,
    "deviceApiUrl": null,
    "jurisdictionSwitch": false,
    "maxNotifyPeriod": [
        
    ],
    "minNotifyPeriod": [
        
    ],
    "redirectUri": null,
    "sid": null,
    "threshold": [
        
    ],
    "threshold_recommended_value": 0,
    "validClientId": null,
    "validClientSecret": null,
    "supportsMultipleDTLS": true
}```

**Comment by Askidea** [comment]
@SiMet Thanks for your kind reply.

In Android, I'm testing my OCF Client app with v2.2.4.3 or master branch and CTT 2103.0.0.

* build options : 
make NDK_HOME=~/toolchains/ndk21-android-arm-23 ANDROID_API=23 DEBUG=1 CLOUD=1 SECURE=1 IPV4=1 TCP=1 PKI=1 DYNAMIC=1 JAVA=1 IDD=1

May I ask for some advice on how to test the next item of CTT?
**(1) CT1.2.6 RETRIEVE Message with observe indication based on CoAP over TCP - NOTIFY**
Prompt: Please change some value in /oic/p resource and press OK 
Prompt: Please change some value in /oic/d resource and press OK 
=====> I tried update /oic/p > mnmn or /oic/d name, but access denied. How can I change /oic/p and /oic/d as prompted?

**(2) CT2.2.7 RETRIEVE Message with observe indication based on CoAP over TCP - NOTIFY**
Prompt: Please have the IUT establish a TCP connection with the CTT.
======> I cannot find a CTT host address. Where can I find CTT address?

If these items are resolved, it is likely that v2.2.4.3 will be verified in the Android environment.
I look forward to your help. 
Thank you in advance.

 My PICS:

`{
    "device": "Device",
    "company": "Askidea",
    "cplAtIANAPen": "",
    "platformVersion": "1.0.0",
    "cplAtVersion": "1.0.0",
    "cplAtModel": "M140",
    "role": "Client",
    "supportedVerticalProfile": [
        "Smart Home"
    ],
    "supportedDeviceTypes": [
        "oic.d.switch",
        "oic.d.light",
	    "oic.d.smartlock"
    ],
    "icv": "ocf.2.2.4",
    "dmv": "ocf.res.1.3.0",
    "acceptVersion": [
        "1.0.0"
    ],
    "resources": [
        "oic.r.csr",
        "oic.r.roles",
        "oic.r.sdi",
        "oic.r.sp",
        "oic.r.coapcloudconf"
    ],
    "optionalFeatures": [],
    "OTM": [
        "oic.sec.doxm.jw"
    ],
    "contentFormatVersion": [
        "1.0.0"
    ],
    "multiValueQuerySupport": false,
    "supportsMultipleDTLS": false,
    "cloudRegisterAutoRetry": false,
    "sct": 9,
    "supportsCoAPoverTCP": true,
    "canTurnOnSoftAPManually": false,
    "observableOICRES": false,
    "supportsCancellingObserve": false,
    "persistentDeviceuuid": false,
    "supportedSecurityProfiles": [
        "sp-baseline-v0"
    ],
    "supportseSIMEasySetup": false
}`

**Comment by SiMet** [comment]
@Askidea 

1) CT1.2.6 in our test app we set /oic/p and /oic/d as not observable. Does your device require to have /oic/d /oic/p resources observable?

2) In test CT2.2.7 CTT starts simulator which you should be able to discover. In discovery response there are endpoints for each resource. You need to look for endpoint which starts with "coap+tcp" or "coaps+tcp" and perform Retrieve operation on such resource.

**Comment by Askidea** [comment]
@SiMet 

(1) CT1.2.6
Regarding Linux, there is the method to set resources(/oic/p, /oic/d, etc)  obsersable or unobserable by getting the resource  by index. But, for Android, probably there is no any java binding method to get resource object about /oic/p or /oic/d.  

[Example: server_certification_tests.c] 
  oc_resource_t *device_resource = **oc_core_get_resource_by_index**(OCF_D, DEVICE);   
  **oc_resource_set_observable**(device_resource, false);     observable);

  oc_resource_t *platform_resource = oc_core_get_resource_by_index(OCF_P, DEVICE); 
  oc_resource_set_observable(platform_resource, false);
-------------
**oc_core_get_resource_by_index**()   ==> No java binding method
**oc_resource_set_observable**() ==> OCMain.resourceSetObservable(OCResource res, boolean 

(2) CT2.2.7
Using the 'client_certification_tests' binary in Linux, it was successful to make a TCP connection by performing a GET(coaps+tcp) request for a resource. However, an OC_STATUS_SERVICE_UNAVAILABLE error occurs when a GET request is performed in an Android app. The resource endpoint of the CTT device simulator seems to have been specified correctly.
I'm still checking on this issue.

**Comment by SiMet** [comment]
I'm not expert in JNI and Swig but there is probably method in Java **OCCoreRes.getResourceByIndex** which is bound to **oc_core_get_resource_by_index**?

Please see: https://github.com/iotivity/iotivity-lite/blob/master/swig/swig_interfaces/oc_core_res.i

**Comment by Askidea** [comment]
@SiMet

Thanks for your advice. I modified the code based on your advice, and CT1.2.6 passed.
For your information, I used a more intuitive **OCCoreRes.getResourceByUri()** instead of getResourceByIndex(). Finding resources with indexes was not easy.

**Comment by SiMet** [comment]
I'm not familiar with Android implementation. Maybe @Danielius1922 could help us or point someone who could help us.

**Comment by Askidea** [comment]
@SiMet, @Danielius1922 
For CT2.2.7, after changing to IPv4/Secure/TCP, the TCP connection was successful and the CT item was passed. But for IPv6, the TCP connection failed as I said above.
Why is IPv6 TCP connection working on Linux and not on Android? Could it be a problem with the mbedtls library?
For now, this is what I've checked.

![image](https://user-images.githubusercontent.com/74634317/193417722-c6f111e8-1593-4057-b190-45e74236696c.png)

**Comment by Danielius1922** [comment]
Sorry, I didn't notice this. I'm not really familiar with Android implementation either, however as far as I'm aware it's almost identical to the Linux v2.2.5.1. 

Your log analysis seems to be on the right track, however I think the OC_CLIENT ifdef is working because the log

```
Posting INIT_TLS_CONN_EVENT
```

is always under OC_CLIENT ifdef, and that log occurs in both of the log traces you posted, so that makes me think it is in order.  So I think what is failing is this:

```C
      oc_sec_cred_t *cred =
        oc_sec_find_creds_for_subject(&endpoint->di, NULL, endpoint->device);
      if (cred && cred->credtype == OC_CREDTYPE_PSK) {
```

It cannot find the PSK certificate. If you look in the log trace of the failing case there is this:
```
I/../../security/oc_tls.c <oc_tls_load_mfg_cert_chain:1229>: loading manufacturer cert chain
I/../../security/oc_tls.c <oc_tls_configure_end_entity_cert_chain:1219>: error configuring identity cert
```

Whilst in the successful case the loading of the manufacturer cert chain succeeds. We need to find out why the loading of manufacturer cert chain is failing.

**Comment by Askidea** [comment]
@Danielius1922 

For IPv4, also following line appeared. But TCP connection initiated successfully (CT 2.2.7).
`
2022-10-12 13:43:17.478 14638-14662 I/../../security/oc_tls.c <oc_tls_load_identity_cert_chain:1238>: loading identity cert chain
2022-10-12 13:43:17.478 14638-14662 I/../../security/oc_tls.c <oc_tls_configure_end_entity_cert_chain:1219>: error configuring identity cert
2022-10-12 13:43:17.478 14638-14662 I/../../security/oc_tls.c <oc_tls_load_mfg_cert_chain:1229>: loading manufacturer cert chain
2022-10-12 13:43:17.478 14638-14662 I/../../security/oc_tls.c <oc_tls_configure_end_entity_cert_chain:1219>: error configuring identity cert
`

Please refer to log files belows.

[CTT2103_CT2.2.7_master_29ccbf39_IPv4_OCFClient logcat_20221012.txt](https://github.com/iotivity/iotivity-lite/files/9761646/CTT2103_CT2.2.7_master_29ccbf39_IPv4_OCFClient.logcat_20221012.txt)

[CTT2103_CT2.2.7_master_29ccbf39_IPv6_OCFClient logcat_20221012.txt](https://github.com/iotivity/iotivity-lite/files/9761648/CTT2103_CT2.2.7_master_29ccbf39_IPv6_OCFClient.logcat_20221012.txt)

[CTT2103_CT2.2.7_master_29ccbf39_IPv6_CTT log_20221012.txt](https://github.com/iotivity/iotivity-lite/files/9761650/CTT2103_CT2.2.7_master_29ccbf39_IPv6_CTT.log_20221012.txt)

**Comment by Danielius1922** [comment]
@Askidea can you please retry with the branch from this PR - https://github.com/iotivity/iotivity-lite/pull/327 ? I've added some logs, lets see if they tell us something.

**Comment by Askidea** [comment]
@Danielius1922 

Here are the log files.

[OCFClient_log_android-tcp-logs_20221014.txt](https://github.com/iotivity/iotivity-lite/files/9782435/OCFClient_log_android-tcp-logs_20221014.txt)

[CTT_log_android-tcp-logs_20221014.txt](https://github.com/iotivity/iotivity-lite/files/9782437/CTT_log_android-tcp-logs_20221014.txt)

And there are compile errors in the branch 'android-tcp-logs' because of a typo, e.g. "ernno".

**Comment by Danielius1922** [comment]
@Askidea 

Sorry about that, I've added the logs but compiled the code without enabling debug logs. I've fixed it in the branch.

You've figured it out anyway, so looking at the logs:

```
2022-10-14 13:02:16.641 29917-29945 I/OC-JNI: Outgoing message of size 110 bytes to 
2022-10-14 13:02:16.641 29917-29945 I/OC-JNI: coaps+tcp://[fe80:0000:0000:0000:2c3d:ce34:1b17:3760]:49176
2022-10-14 13:02:16.641 29917-29945 I/OC-JNI: 
2022-10-14 13:02:16.641 29917-29945 I/OC-JNI: could not find ongoing TCP session for endpoint:
2022-10-14 13:02:16.641 29917-29945 I/OC-JNI: coaps+tcp://[fe80:0000:0000:0000:2c3d:ce34:1b17:3760]:49176
2022-10-14 13:02:16.641 29917-29945 I/OC-JNI: 
2022-10-14 13:02:16.645 29917-29945 I/tcpadapter.c <connect_nonb:494>: failed to connect to address (error=22)
2022-10-14 13:02:16.645 29917-29945 I/tcpadapter.c <initiate_new_session:567>: connect failed, retry(1)
2022-10-14 13:02:16.651 29917-29945 I/tcpadapter.c <connect_nonb:494>: failed to connect to address (error=22)
2022-10-14 13:02:16.651 29917-29945 I/tcpadapter.c <initiate_new_session:567>: connect failed, retry(2)
2022-10-14 13:02:16.655 29917-29945 I/tcpadapter.c <connect_nonb:494>: failed to connect to address (error=22)
2022-10-14 13:02:16.655 29917-29945 I/tcpadapter.c <initiate_new_session:567>: connect failed, retry(3)
2022-10-14 13:02:16.658 29917-29945 I/tcpadapter.c <connect_nonb:494>: failed to connect to address (error=22)
2022-10-14 13:02:16.658 29917-29945 I/tcpadapter.c <initiate_new_session:567>: connect failed, retry(4)
2022-10-14 13:02:16.660 29917-29945 I/tcpadapter.c <connect_nonb:494>: failed to connect to address (error=22)
2022-10-14 13:02:16.660 29917-29945 I/tcpadapter.c <initiate_new_session:567>: connect failed, retry(5)
2022-10-14 13:02:16.660 29917-29945 I/tcpadapter.c <initiate_new_session:571>: could not initiate TCP connection
2022-10-14 13:02:16.660 29917-29945 I/tcpadapter.c <oc_tcp_send_buffer:611>: could not initiate new TCP session
```
Error 22 is `EINVAL` = invalid input. The failing call looks like this:
```
if ((n = connect(sockfd, (struct sockaddr *)r, r_len)) < 0) {
```

`sockfd` must be valid, because the `fcntl` calls preceding the `connect` call succeed. The problematic value must be then the `r` value with the socket address. The address used (fe80::2c3d:ce34:1b17:3760) is a link-local address and as I understand it for link-local addresses sin6_scope_id must be set to >0 value. Unfortunately, the logging functions ignore the scope_id value, so we don't know whether the value was set or not, but given the fact that we get `EINVAL` error it must not be set. 

Now it's possible that the scope_id value gets zeroed somewhere or it was perhaps never set?

At the beginning of the logs I see the following lines
```
2022-10-14 13:02:16.628 29917-29917 E/OCFClientService: (OCF-Client -----------------------------------------------------> MPBridge)
2022-10-14 13:02:16.631 29917-29917 I/EpUtil: ep[0] = coaps://[fe80::2c3d:ce34:1b17:3760]:49175
2022-10-14 13:02:16.631 29917-29917 I/EpUtil: ep[1] = coaps+tcp://[fe80::2c3d:ce34:1b17:3760]:49176
2022-10-14 13:02:16.632 29917-29917 I/iotivity-lite-java/jni/oc_endpoint_wrap.c <jni_string_to_endpoint_a:310>: JNI: jni_string_to_endpoint_a
2022-10-14 13:02:16.634 29917-29917 E/OCFClientService: doGet() : dt=oic.d.light, uri=/BinarySwitchResURI, query=[rt=oic.r.switch.binary],  coaps+tcp://[fe80::2c3d:ce34:1b17:3760]:49176
```

I see two link-local addresses there, but no scope_id, though I don't know if the logging functions just ignore it or it is indeed not set. What are you using to print the addresses? And where do the addresses come from, some sort of configuration? If it is 
 a configuration somewhere, then maybe it is missing the scope_id value in the address...?

**Comment by Askidea** [comment]
@Danielius1922 
I found the cause of the problem from your advice.
The cause was that the OCEndpoint object was converted to an endpoint string, and this endpoint string was converted back to an OCEndpoint object. [ by using OCEndpointUtil.toString(OCEndpoint ep) and OCEndpointUtil.stringToEndpoint(String epString ]

OCEndpoint object --------- OCEndpointUtil.toString(OCEndpoint ep) ----------> endpoint string (e.g. coaps+tcp://[fe80::2c3d:ce34:1b17:3760]:49176) 
---------- OCEndpointUtil.stringToEndpoint(String epString) ----------> new OCEndpoint object

I think the scopo_id information you mentioned was lost in the process of converting the endpoint string into an OCEndpoint object.

In conclusion, CTT2.2.7 passed well in IPv6/Secure/TCP.
As such, this issue seems to be closed, and I am very grateful for your hard work and response.

**Comment by Danielius1922** [comment]
Great to know! Glad I was able to help. :-)

(Before closing this issue I'll first investigate if the conversion from endpoint to string and back preserves the scope_id. I think that would constitute a bug.)

**Comment by Danielius1922** [comment]
Indeed when converting from a string the scope_id value is not filled (and also the parsing function `oc_string_to_endpoint` will be stuck in an infinite loop when an ipv6 address with scope id is used as an input). I've created an enhancement to fix it - https://github.com/iotivity/iotivity-lite/issues/331 .
